File: typescript/provenance-mark/src/rng-state.ts
/**
 * RNG state -- a 32-byte wrapper for the Xoshiro256** state.
 */

import { type Cbor, toByteString, expectBytes } from '@bc/dcbor';
import { ProvenanceMarkError } from './error.js';
import { bytesToHex, bytesEqual, toBase64, fromBase64 } from './utils.js';

export const RNG_STATE_LENGTH = 32;

export class RngState {
  readonly #data: Uint8Array;

  private constructor(data: Uint8Array) {
    this.#data = data;
  }

  toBytes(): Uint8Array {
    return new Uint8Array(this.#data);
  }

  static fromBytes(bytes: Uint8Array): RngState {
    if (bytes.length !== RNG_STATE_LENGTH) {
      throw new ProvenanceMarkError(
        'InvalidSeedLength',
        `invalid RNG state length: expected ${RNG_STATE_LENGTH} bytes, got ${bytes.length} bytes`,
      );
    }
    return new RngState(new Uint8Array(bytes));
  }

  hex(): string {
    return bytesToHex(this.#data);
  }

  toCbor(): Cbor {
    return toByteString(this.#data);
  }

  static fromCbor(c: Cbor): RngState {
    const bytes = expectBytes(c);
    return RngState.fromBytes(new Uint8Array(bytes));
  }

  toJSON(): string {
    return toBase64(this.#data);
  }

  static fromJSON(s: string): RngState {
    const bytes = fromBase64(s);
    return RngState.fromBytes(bytes);
  }

  equals(other: RngState): boolean {
    return bytesEqual(this.#data, other.#data);
  }
}

File: python/provenance-mark/src/provenance_mark/_rng_state.py
"""PRNG state wrapper."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from dcbor import CBOR

RNG_STATE_LENGTH = 32


@dataclass(frozen=True, slots=True)
class RngState:
    """A 32-byte serialized Xoshiro256** state."""

    _data: bytes

    def __post_init__(self) -> None:
        if len(self._data) != RNG_STATE_LENGTH:
            raise ValueError(
                (
                    "invalid RNG state length: expected 32 bytes, "
                    f"got {len(self._data)} bytes"
                )
            )

    @staticmethod
    def from_bytes(data: bytes | bytearray) -> RngState:
        return RngState(bytes(data))

    @staticmethod
    def from_slice(data: bytes | bytearray) -> RngState:
        return RngState.from_bytes(data)

    def to_bytes(self) -> bytes:
        return bytes(self._data)

    def hex(self) -> str:
        return self._data.hex()

    def to_cbor(self) -> CBOR:
        return CBOR.from_bytes(self._data)

    def to_cbor_data(self) -> bytes:
        return self.to_cbor().to_cbor_data()

    @staticmethod
    def from_cbor(cbor: CBOR) -> RngState:
        return RngState.from_bytes(cbor.try_byte_string())

    def to_json(self) -> str:
        import base64

        return base64.b64encode(self._data).decode("ascii")

    @staticmethod
    def from_json(value: str | dict[str, Any]) -> RngState:
        import base64

        if isinstance(value, dict):
            raise TypeError("RngState JSON representation is a string")
        return RngState.from_bytes(base64.b64decode(value, validate=True))

